lib/match.rb

```diff
@@ -0,0 +1,150 @@
+require_relative "board"
+require_relative "player"
+
+class Match
+
+
+    def initialize(board, player_1, player_2)
+
+        @board = board 
+        @player_1 = player_1
+        @player_2 = player_2
+
+        @turn_type = decide_turns
+        @moves = []
+
+        @winner_type = ""
+        @board.welcome_message(@player_1.token, @player_2.token)
+        
+    end
+
```

Comment:
The `Match` class is still doing too many things. To have a high level idea of what `Match`'s responsibility is, we can state that "`Match` is responsible for making each player do its turn while the game is not yet finished". If you take a look at the code snippet I sent in my previous comment/annotation above, the skeleton/template of `Match` should be simply representing this responsibility.

To make this clear, imagine this. Suppose `Match` is a company/business that you own whose sole operational mission is to make sure each player will make its turn until the game is finished. Now as the CEO, you don't want your company to have lots of machines (methods) because when those machines break down `Match` will be responsible in fixing them. As CEO of this company `Match`, you want to own as few machines (methods) as possible to do your company's mission. With that those other machines that `Match` needs but don't want to operate can be subcontracted into a type of business called a `module`. 

For example, `check_length`. Probably `Match` doesn't have to own this machine (method). Probably some other business (module) should own this method and `Match` can simply use its services here (by simply invoking from that module this `check_length` method when needed). 

Anyway I hope that metaphor can help us better visualize we can refactor `Match` further. When confused you can always reach out to me :) Just please consider the snippet/template I provided you guys previously. Try as much as possible to limit your `Match` 's methods to those that I showed in that snippet. Every other method should then just belong to outside modules that `Match` will include.

Reply:
I see. It's a good analogy. We decided to put two modules: validation and turn management. The validations is taking care of the input; checking lengths and availability. This will help the Match class to just ask if the input that the player class is validated by the validation module. I think this is a great implementation idea. Also, the turn management is here to help the Match class to setup the initial turns, setting the next turn, and so on, so now Match just asks for it instead of taking charge. Great idea also.

Now, we take out the moves array from match, as your suggestion implied. We put the moves array in each player class. So when the validation module takes a move, it will ask each player class which values have they store, in order to know if the actual move is valid or not.